pkg/controller/sparkapplication/controller.go

```diff
@@ -607,37 +602,55 @@ func (c *Controller) recordExecutorEvent(
 	}
 }
 
-// shouldSubmit determines if a given application is subject to a submission retry.
+// shouldSubmit determines if a given application is subject to a retry.
 func shouldSubmit(app *v1alpha1.SparkApplication) bool {
-	// App was never attempted. Should submit.
-	if app.Status.Attempts == 0 {
-		return true
-	}
 
-	// No retries specified or all retries exhaused.
-	if app.Spec.FailureRetries == nil ||
-		app.Status.Attempts >= (*app.Spec.FailureRetries)+1 {
+	switch app.Status.AppState.State {
+	case v1alpha1.NewState:
+		return true
+	case v1alpha1.SubmittedState, v1alpha1.RunningState:
 		return false
-	}
+	case v1alpha1.CompletedState:
+		if app.Spec.RestartPolicy.Type == v1alpha1.Always {
+			return true
+		}
+	case v1alpha1.FailedState:
+		shouldRetry := false
+		if app.Spec.RestartPolicy.Type == v1alpha1.Always {
+			shouldRetry = true
+		} else if app.Spec.RestartPolicy.Type == v1alpha1.OnFailure {
+			// We Retry if we haven't hit the retry limit.
+			if app.Spec.RestartPolicy.OnFailureRetries != nil && app.Status.Attempts < (*app.Spec.RestartPolicy.OnFailureRetries)+1 {
+				shouldRetry = true
+			}
+		}
 
-	// Retries not exhaused and no Interval Specified.
-	if app.Spec.RetryInterval == nil {
-		return true
-	}
+		if shouldRetry && app.Spec.RestartPolicy.OnFailureRetryInterval != nil {
+			// Retry if we have waited at-least equal to attempts*RetryInterval since we do a linear back-off.
+			interval := time.Duration(*app.Spec.RestartPolicy.OnFailureRetryInterval) * time.Second * time.Duration(app.Status.Attempts)
+			currentTime := time.Now()
+			return currentTime.After(app.Status.CompletionTime.Time.Add(interval))
+		}
 
-	// Retry if we have waited atleast equal to RetryInterval.
-	interval := time.Duration(*app.Spec.RetryInterval) * time.Second
-	currentTime := time.Now()
+	case v1alpha1.FailedSubmissionState:
+		shouldRetry := false
+		if app.Spec.RestartPolicy.Type == v1alpha1.Always {
+			shouldRetry = true
+		} else if app.Spec.RestartPolicy.Type == v1alpha1.OnFailure {
+			// We Retry if we haven't hit the retry limit.
+			if app.Spec.RestartPolicy.OnSubmissionFailureRetries != nil && app.Status.SubmissionAttempts < (*app.Spec.RestartPolicy.OnSubmissionFailureRetries)+1 {
+				shouldRetry = true
+			}
+		}
 
-	var lastEventTime metav1.Time
-	if !app.Status.CompletionTime.IsZero() {
-		lastEventTime = app.Status.CompletionTime
-	} else {
-		lastEventTime = app.Status.SubmissionTime
-	}
-	if currentTime.After(lastEventTime.Time.Add(interval)) {
-		return true
+		if shouldRetry && app.Spec.RestartPolicy.OnSubmissionFailureRetryInterval != nil {
+			// Retry if we have waited at-least equal to attempts*RetryInterval since we do a linear back-off.
```

Comment:
we do update the SubmissionTime or CompletionTime based on the last event. 

attempts* retryInterval is because we do a linear backoff .. so if our interval is 1 sec and this is 3rd retry, our backoff time is 3 * 1 = 3 seconds ... similarly for 4th retry, it will be 4 seconds and so on